add clang-tidy support

JohnCoconut opened the pull request:
Only two options are enabled (hard-coded in script)

* performance-faster-string-find
* performance-for-range-copy

Treat warnings as errors are enabled.

Another try.

JohnCoconut committed:
add clang-tidy support

JohnCoconut commented:
YAML:1:1: error: not a mapping
     /home/travis/build/cocos2d/cocos2d-x/tools/travis-scripts/../../.clang-tidy

Seems to complain my `.clang-tidy` is invalid, but it runs fine in my local linux machine. Is it because I am use clang 8?

The format is copied from [CMake](https://github.com/Kitware/CMake/blob/07cfb18f9d29cfc0588ede928846a03ec5599c48/.clang-tidy) git repo.

JohnCoconut committed:
move clang_tidy_test up in travis yaml file

so that it's run first

JohnCoconut committed:
use dump-config to generate .clang-tidy file

minggo commented on tools/travis-scripts/run-script.sh:
```diff
@@ -37,6 +37,16 @@ function build_linux()
+    mkdir -p clang-tidy-build
+    cd clang-tidy-build
+    cmake ../.. -DCMAKE_EXPORT_COMPILE_COMMANDS=on
+    python /usr/local/clang-7.0.0/share/clang/run-clang-tidy.py
```
```
cmake ../.. -DCMAKE_EXPORT_COMPILE_COMMANDS=on
python /usr/local/clang-7.0.0/share/clang/run-clang-tidy.py
```

Don't quite understand these codes. And the python script path is hard coded, it may be a problem when CI changes its environment.

JohnCoconut replied:
I `apt search clang-tidy` in travis ci and found the most recent version is 6. But the `clang-tidy` being used is 7. My guess is that `clang-tidy` is not installed by package manager. I am not sure how to find `run-clang-tidy.py`.

Or we can just copy one and store it in the repo, if you agree.

crazyhappygame replied:
@minggo 
`cmake ../.. -DCMAKE_EXPORT_COMPILE_COMMANDS=on`
CMAKE_EXPORT_COMPILE_COMMANDS is used to generate compilation database. This is the easiest way to use clang-tidy with CMake 
https://clang.llvm.org/docs/JSONCompilationDatabase.html

About hard-coded path 
`python /usr/local/clang-7.0.0/share/clang/run-clang-tidy.py`
It can be changed to e.g.:
"python $(dirname $(dirname $(readlink -f \`which clang-tidy\`)))/share/clang/run-clang-tidy.py"

JohnCoconut replied:
@crazyhappygame Great. It works on Fedora Linux too. I didn't know clang-tidy shares top-level directory with its python script.

minggo replied:
@crazyhappygame thanks for the explanation.

JohnCoconut committed:
use flexible way to find run-clang-tidy.py

clang-tidy and run-clang-tidy.py shares common ancester directory.
find the script dynamically rather than hard-coding it.

crazyhappygame commented on .clang-tidy:
```diff
@@ -0,0 +1,38 @@
+---
+Checks:          'clang-diagnostic-*,clang-analyzer-*,-*,performance-faster-string-find,performance-for-range-copy,'
```
Please change to yaml multplie line string https://yaml-multiline.info/. In that way we can easily track changes
```
Checks: >
    -*,
    clang-analyzer-*,
    clang-diagnostic-*,
    performance-faster-string-find,
    performance-for-range-copy,
```

JohnCoconut replied:
```yaml
Checks:          'clang-diagnostic-*,clang-analyzer-*,-*,performance-faster-string-find,performance-for-range-copy,'
```

The configuration above is generated by `clang-tidy -header-filter='/(?!external)/.*' -checks='-*,performance-faster-string-find,performance-for-range-copy,' -warnings-as-errors='*' -dump-config`


```yaml
Checks: >
    -*,
    clang-analyzer-*,
    clang-diagnostic-*,
    performance-faster-string-find,
    performance-for-range-copy,
```

I just found that the above two are not the same. The order matters. `-*` will disable options ahead of it.

* In the first config, `clang-diagnostic-*,clang-analyzer-*` is disabled.

* The second config, placing `-*` at the front, will not disable `clang-diagnostic-*,clang-analyzer-*`.

JohnCoconut committed:
place clang-tidy check options in separate lines

crazyhappygame commented on tools/travis-scripts/run-script.sh:
```diff
@@ -291,6 +301,12 @@ function run_pull_request()
 
+    # linux_clang_tidy_test
+    if [ $BUILD_TARGET == 'linux_clang_tidy' ]; then
+        genernate_binding_codes
```
I think that genernate_binding_codes should be removed

minggo replied:
Yep, it is not needed.

crazyhappygame commented on .clang-tidy:
```diff
@@ -0,0 +1,44 @@
+---
+Checks: >
```
Why  "-*" is needed?
I thought that plan was enable "performance-faster-string-find", "performance-for-range-copy".
```
Checks: >
  performance-faster-string-find,
  performance-for-range-copy,
```
Are you sure that all "clang-diagnostic-*" , "clang-analyzer-*" checks are already fixed?

JohnCoconut replied:
@crazyhappygame  `clang-tidy` enables a lot of default options. The `-*` option is to remove all options currently enabled.

crazyhappygame commented on .clang-tidy:
```diff
@@ -0,0 +1,44 @@
+HeaderFilterRegex: '/(?!external)/.*'
+AnalyzeTemporaryDtors: false
+FormatStyle:     none
+CheckOptions:    
```
I propose to remove "CheckOptions:".
Any way default values for "CheckOptions:" were used

JohnCoconut committed:
clang_tidy_test target clean up

crazyhappygame commented:
@minggo could you merge it?

minggo commented:
Sure, it is merged.
